Add PR advice

eloch216 opened the pull request:
This PR adds a new section to the developer documentation that describes a few conventions we tend to follow when making and reviewing PRs. Hopefully this will be useful for people who are getting started with contributing to BioCro.

@leavesandlemmas As the newest contributor, it would be great to get your thoughts on this -- do you think it is clear and helpful?

leavesandlemmas commented on developer_documentation/contribution_guidelines.Rmd:
```diff
@@ -176,6 +176,85 @@ the devil].
+  treat each branch and PR as having a single "owner" who is in charge of making
+  all commits. There are a few consequences of this approach:
+
+  - Pull request reviewers should only suggest changes, not implement them. If a
```
What about suggested implementations?

justinmcgrath replied:
Maybe this "Pull request reviewers should suggest changes, not commit changes directly to the PR branch."

eloch216 replied:
I'll use Justin's suggested version. Scott, does that clear up your question?

leavesandlemmas replied:
Yes

leavesandlemmas commented on developer_documentation/contribution_guidelines.Rmd:
```diff
@@ -176,6 +176,85 @@ the devil].
+  for any code to be incorporated into BioCro. The following is a thorough
+  workflow that should ensure submitted code is working:
+
+  1. Build the package and run the tests locally. The tests don't take much time
```
Build or install ?  `R CMD install biocro` or `R CMD build biocro` ?

eloch216 replied:
I'll change this to just "run the tests locally" and link to the instructions for doing so.

leavesandlemmas commented on developer_documentation/contribution_guidelines.Rmd:
```diff
@@ -176,6 +176,85 @@ the devil].
+    responsibility of the pull request owner to finalize the PR by merging it
+    and deleting the branch.
+
+* **Submitted code should work:** It takes time and effort to produce a
```
A command line walkthough would be helpful here, or example code. Or maybe link to the developer manual ?

eloch216 replied:
I thought we had instructions for running `R CMD check`, but I guess we don't. I added some in a dedicated section and put a link here.

leavesandlemmas commented on developer_documentation/contribution_guidelines.Rmd:
```diff
@@ -176,6 +176,85 @@ the devil].
+     R. Sometimes an issue will appear that didn't occur locally on your own
+     operating system / R version.
+
+  4. When the online checks have passed, then assign reviewers to begin the
```
What about dealing with problems in a certain implementation?  Justin didn't like a workaround I created to pass a test, and I undid some work to remove the work around.  How should someone share code that doesn't work or doesn't pass tests?  Make a commit and then ask for review on the commit, rather than a pull request?

eloch216 replied:
Such a PR could also be marked as a "draft PR"; I added that to the advice.

yufenghe22 replied:
I often do `R CMD install .` and then `Rscript testthat.R`. Is it the same as R CMD check?

eloch216 replied:
> I often do `R CMD install .` and then `Rscript testthat.R`. Is it the same as R CMD check?

It's not the same. The `testthat.R` script only runs the regression tests in the `tests` directory. Runing `R CMD check` will run those tests, but will also build the vignettes, run the examples in the help files, check help files for any syntax errors, and a bunch of other stuff. I can try to clarify this in the documentation.

justinmcgrath reviewed (approved):
I think it looks good and agree that the details about which commands (R CMD INSTALL, R CMD build) are intended in each of the places would help people who are not yet familiar with the terminology.

eloch216 commented:
I added some links to instructions for locally running the tests and `R CMD check`. The `R CMD check` instructions are new so they should also be checked for clarity.

The links themselves cannot be followed in the raw markdown documents. To see them in action, it's necessary to build the bookdown documentation. This can be done by calling `bookdown::render_book()` from an R session running in `biocro/bookdown`. There are also instructions in the developer manual itself: https://biocro.github.io/BioCro-documentation/latest/bookdown/generating-documentation.html#compiling-and-viewing-the-bookdown-book.

leavesandlemmas commented on developer_documentation/running_r_cmd_check.Rmd:
````diff
@@ -0,0 +1,87 @@
+package can be checked with
+
+```
+R CMD build --as-cran BioCro_X.Y.Z.tar.gz
````
Typo? Should be `R CMD check ...`  right?

eloch216 replied:
Whoops, yes it should be!

yufenghe22 commented on developer_documentation/contribution_guidelines.Rmd:
```diff
@@ -176,6 +176,85 @@ the devil].
+  `R CMD check` passes locally.
+
+* If there are `R CMD check` errors but you cannot figure out how to resolve
+  them, a PR could be made to ask for help from other developers.
```
This may be a minor issue. Does it need to have another `PR` to ask for help? What about `Issues` or `Discussions`? Would they be better for posting minor issues?

I think `R CMD check` could also be part of discussions in the same PR, so the owner can make patches according to reviewers' suggestions.

justinmcgrath replied:
There are probably cases where one or the other of this is better, or all are equally fine. It's hard to put everything into the guidelines, but I'd say there's a degree of flexibility with most of them.

eloch216 replied:
Hm. I didn't want this PR suggestion to sound like a requirement. An `Issue` or `Discussion` would also work.

I think the main point I want to convey is that a failing `R CMD check` is fine as long as it is acknowledged. This can be done via a PR, a draft PR, an issue, a discussion, or really anything that seems appropriate. I will think about how to reword this.

justinmcgrath commented on developer_documentation/running_r_cmd_check.Rmd:
```diff
@@ -0,0 +1,87 @@
+R includes a command-line tool called `R CMD check` whose purpose is to check
+whether an R package meets key requirements. At minimum, any package submitted
+to CRAN must pass `R CMD check` without producing any errors. For more
+information about the individual checks that comprise `R CMD check`, please see
```
I understand that few people in science seem to use the word this way, but "comprise" is a near synonym for "include", specifically it means "includes in its entirety". Whenever I read something like "the parts comprise the whole" it takes a second to realize that "compose" was intended.

eloch216 replied:
I didn't know that! 🤯

justinmcgrath replied:
I did look it up to see where other people fell in this. TFD indicates that still a slight majority don't like for it to mean compose: https://www.thefreedictionary.com/comprise.

The description at Merriam Webster indicates that they think the distinction isn't useful, saying it's not clear why anyone would single out "compose" as wrong. To me it's clear enough that it's confusing for a word to have two opposite sorts of meaning, but maybe I should just give in. https://www.merriam-webster.com/dictionary/comprise

eloch216 replied:
I guess I was never taught this rule and I've never looked it up. I tend to think of both "compose" and "comprise" as describing a relationship between parts and a whole, but usually from the context is clear which noun is a part and which is a whole, so I don't usually differentiate between "compose" and "comprise."

leavesandlemmas replied:
> it's confusing for a word to have two opposite sorts of meaning,

[Contranyms](https://en.wikipedia.org/wiki/Contronym) are pretty common. Some examples are *to sanction* which can mean to approve an action (The university sanctioned the experiment) or to penalize someone for an action. *oversight*  can mean supervision/care or a lack of supervision/care.  When you  dust a house  with a cloth, you remove dust, but when you dust a cake with powdered sugar, you add powdered sugar.

gsrohde replied:
For what it's worth, the Oxford English Dictionary lists, among the definitions of _comprise_, 

> 8.b. To constitute, make up, compose.

and dates this usage as far back as 1794:

> 1794
> The propositions which comprise the several heads of our testimony.
> W. Paley, View of Evidences of Christianity vol. I. i. ix. 212

A more recent usage:

> 1969
> These essays together with those contained in this volume comprise the total of C. S. Lewis's essays on literature.
> W. Hooper in C. S. Lewis, Sel. Lit. Essays p. xix

(Of course, the OED is British!)

I used to be in Justin's camp, believing that "comprise" should only be used to mean "include" or "encompass", but I've since changed my mind.

eloch216 commented:
I added some more information to clarify how `R CMD check` is different from the regression tests, reworded the advice for how to deal with an `R CMD check` error, and fixed a few typos / grammar mistakes.

gsrohde commented on developer_documentation/contribution_guidelines.Rmd:
```diff
@@ -176,6 +177,107 @@ the devil].
+  thoughtful review of a pull request, so requests to review a PR should
+  generally only be made for working code. In a BioCro context, "working" code
+  means that `R CMD check` does not produce any errors. See
+  [Running R CMD check](#r-cmd-check) (Chapter \@ref(r-cmd-check)) for more
```
It seems to me redundant and unnecessary to hyper-link on both "R CMD check" and on "4", but if you _are_ going to link on the latter, I'd include "Chapter" in the link text.

eloch216 replied:
I based this cross-reference off one I found in `tests/README.md`: https://github.com/biocro/biocro/blame/31a414542d64513ba0c5425ee2b72fe651670fe5/tests/README.md#L41-L43

I'm not sure how to get `\@ref` to include the word "Chapter" in the resulting hyperlink, and I don't want to have a hard-coded `4` (in case the chapter numbers change), so I can just get rid of it. It could be something like:

```
See [the "Running R CMD check" chapter](#r-cmd-check) for more
```

gsrohde replied:
Yes, that was my code.<sup>*</sup>

Maybe there _is_ no good way to do what I was thinking should be done without hard-coding the chapter number, which is probably a step backwards.  I used to work for a company that took web accessibility seriously, but that was over a decade ago,  so I haven't kept up to date with current thinking about such issues and don't know if the accessibility issue I brought up is significant enough to even bother with.  (There is, by the way, a group on campus that deals with electronic document accessibility issues and meets regularly and has a campus listserve.  I'm on their e-mail list but haven't gone to any meetings since leaving my job at the afore-mentioned company.)

Your solution seems viable, if you want to bother.  If you do make the change, there are at least two similar cases in this new, added section, and maybe others (in addition to the one you mention) elsewhere in the manual.

<sup>*</sup> I have, maybe even more that once, had the experience of looking at some old code and thinking "Who the heck wrote this crud!" only to discover from the version-management system that that "who" was me.  I suppose the positive spin on this is that it might suggest my coding skills had significantly improved.

gsrohde commented on developer_documentation/contribution_guidelines.Rmd:
```diff
@@ -176,6 +177,107 @@ the devil].
+  discuss it but there are `R CMD check` errors, this can be accomplished via a
+  PR, a draft PR, a GitHub issue, or a GitHub discussion. The best approach may
+  vary with the particular situation. Regardless of which approach is taken, it
+  is essential to acknowledge the the `R CMD check` issue and be clear about the
```
the the -> the

(While there are certainly cases where "that that" may occur in a perfectly acceptable sentence, e.g. "I didn't think that that was possible"—though I have had the experience of having this edited out of my own writing, much to my annoyance (I blamed the probable bone-headed use by the editor of a grammar checker)—I can't think of similar valid case for "the the" except perhaps with some additional punctuation (e.g., 'The "the" was extraneous.').
